### ISSUE TITLE
Firebase does not accept messages when a custom JsonSerializerSettings is provided

### ISSUE THREAD

**Issue by sir-boformer** [bytefish/FcmSharp#35]
I'm using FcmSharp in Asp .Net Core. I Created a `JsonSerializer` based on the default `JsonSerializerSettings` provided by ASP .Net Core MVC so that things like date formatting and support for custom serializers are applied to notification payloads.

When I sent a message with a simple `ApnsConfig` (just an alert with title and body), this error occured:

```
FcmSharp.Exceptions.FcmMessageException: {
  "error": {
    "code": 400,
    "message": "Request contains an invalid argument.",
    "errors": [
      {
        "message": "Request contains an invalid argument.",
        "domain": "global",
        "reason": "badRequest"
      }
    ],
    "status": "INVALID_ARGUMENT"
  }
}
```

It turned out that the json serializer was including null values in the serialized messages. Firebase does not accept a `ApnsConfig` with null properties.

When I set `NullValueHandling = NullValueHandling.Ignore` in the `JsonSerializerSettings`, everything works as expected.

I think it doesn't make sense to allow the creation of a `FcmClient` with an incompatible `JsonSerializer`. 

It would be nice to control the serialization of the `CustomData` dictionary values of the `ApnsConfigPayload` (e.g. use snake_case, date formatting), but it should not affect the serialization of the rest of the message.

**Comment by bytefish** [comment]
This is totally correct! 

I think the default ``FcmSharp`` serializer already uses ``NullValueHandling.Ignore`` as default:

![image](https://user-images.githubusercontent.com/239373/46918383-ef050c00-cfd1-11e8-85a4-f2443f3025eb.png)

What about the following: I make the constructor ``protected``, so you can only obtain an instance using the static ``Default`` method. And if one needs additional options, then the JsonSerializer needs to be overriden.

Does that sound OK?

**Comment by bytefish** [comment]
Ah, I think I get the full picture now. You need to customize the serialization of the ``CustomData``, while the rest of the message needs to be serialized with the default Serializer. Well, I think the ``CustomData`` in the messages is only serialized with the default serializer, if you pass complex objects into it, which in turn will be serialized with JSON.NET again.

If you pass strings in it, then I think JSON.NET won't touch them. There is actually a Unit Test, that shows it:

![image](https://user-images.githubusercontent.com/239373/46918817-07c3f080-cfd7-11e8-9e92-e1dc5fd364ac.png)

I know it isn't really convenient API-wise, if you have to first serialize the values to a ``Dictionary<string, string>`` yourself, but I cannot think of a clean API to provide ``CustomData``-specific serialization / deserialization.

**Comment by sir-boformer** [comment]
It's not really a problem if you know it, I just wanted to report it.

I wanted actual json in the payload, so I just converter my objects into a `JObject` using my own serializer:

```c#
var somePayload = new
{
    testNumber = 1,
    testDate = DateTime.Now,
    testString = "Hello"
};

...

CustomData = new Dictionary<string, object>
{
    ["test"] = JObject.FromObject(somePayload, _jsonSerializer)
}
```

This results in a nice custom payload (for testing purposes I used snake_case):

```json
{
   "validate_only":false,
   "message":{
      "apns":{
         "payload":{
            ...
            "test":{
               "test_number":1,
               "test_date":"2018-10-14T20:45:42.6582191Z",
               "test_string":"Hello"
            }
         }
      },
      "token":"..."
   }
}
```

**Comment by bytefish** [comment]
Oh I forgot, did you try using the Attributes like ``JsonProperty`` on the Properties? That should work just as well and maybe you can also attribute the class and set class-specific Json.NET attributes there.

**Comment by bytefish** [comment]
A tutorial would be great, I am currently having another open issue #34, which is about creating more examples. We could add it to the Examples Folder, where I have already prepared an Android client, too.

**Comment by sir-boformer** [comment]
Yep, although that doesn't work for more advanced stuff like date formatting or custom serialization. Also I think it's a bad idea to use `JsonProperty` if you only want to use a different notation like snake_case. It's just another error source.

I think this issue can be closed.
